### DISCUSSION TITLE
SPI peripheral usage from ISR in light sleep

### DISCUSSION THREAD

**Discussion by dalbert2** [espressif/arduino-esp32#8919] [Q&A]
I am using the SPIClass class to interface with a radio transceiver.  Timing is critical and certain things must be handled in the interrupt context: for example, when the radio starts receiving, it generates an interrupt on a GPIO pin.  I decorate all of my functions that run in the interrupt context with IRAM_ATTR and everything works great...until I enable light_sleep.

``` 
  esp_sleep_enable_timer_wakeup(5*1000ULL);  // wake every 5s for housekeeping
  uart_set_wakeup_threshold(UART_NUM_0, 3); // wake on 3 edges of UART activity
  esp_sleep_enable_uart_wakeup(UART_NUM_0);
  esp_sleep_pd_config(ESP_PD_DOMAIN_VDDSDIO, ESP_PD_OPTION_ON); // keep UART2 pins active
  esp_sleep_enable_ext1_wakeup((1ULL<<GPIO_NUM_25) |  // DIO4/Preamble
                               (1ULL<<GPIO_NUM_35) |  // DIO2/SynAddr
                               (1ULL<<GPIO_NUM_26) |  // DIO0/Packet
                               (1ULL<<GPIO_NUM_34),   // Coprocessor
                               ESP_EXT1_WAKEUP_ANY_HIGH);
  esp_light_sleep_start();
  // restore normal operation of GPIOs used for external wakeup
  rtc_gpio_deinit(GPIO_NUM_25);
  rtc_gpio_deinit(GPIO_NUM_35);  
  rtc_gpio_deinit(GPIO_NUM_26);
  rtc_gpio_deinit(GPIO_NUM_34);  
```

With light sleep enabled, I frequently get crashes like:
`rst:0x8 (TG1WDT_SYS_RESET)
`
or
`rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
`
or
```
Guru Meditation Error: Core  1 panic'ed (Cache disabled but cached memory region accessed).

Core  1 register dump:
PC      : 0x400e010c  PS      : 0x00060035  A0      : 0x800820f4  A1      : 0x3ffbfd2c
A2      : 0x0000003f  A3      : 0x3ffc55b8  A4      : 0x80084cac  A5      : 0x3ffb7870  
A6      : 0x3ff000dc  A7      : 0x00000001  A8      : 0x80081ac0  A9      : 0x00000007
A10     : 0x3ffc55b8  A11     : 0x00989680  A12     : 0x00000001  A13     : 0x3ffb7850  
A14     : 0x3ffc6e7c  A15     : 0x3ffc6e7c  SAR     : 0x00000006  EXCCAUSE: 0x00000007
EXCVADDR: 0x00000000  LBEG    : 0x4008b66c  LEND    : 0x4008b677  LCOUNT  : 0xffffffff  


Backtrace: 0x400e0109:0x3ffbfd2c |<-CORRUPTED

  #0  0x400e0109:0x3ffbfd2c in SPIClass::~SPIClass() at C:/Users/david/.platformio/packages/framework-arduinoespressif32/libraries/SPI/src/SPI.cpp:65
```
However, with light sleep disabled, the code runs without any problems and interrupt-driven SPI access works perfectly.

Any suggestions for how to get interrupt-driven access to the SPI bus working in conjunction with light sleep mode would be much appreciated!

**Comment by me-no-dev** [comment]
You should use FreeRTOS Queue, which you will fill with events from the interrupt and handle them in a task that is running with high priority. Chances are, you will be handling the interrupt in 2-3 microseconds. No dangerous code in ISR.

**Reply by dalbert2** [reply]
The ISR code is not "dangerous"; it is placed in RAM, makes no blocking calls, and only a few SPI operations are transacted at 10MHz so the ISR finishes in less than 10uS; most importantly, it works fine unless light sleep is enabled and I would like to understand *why*.

**Reply by me-no-dev** [reply]
Are you using Arduino's SPI class? Because SPI uses FreeRTOS semaphore that is not 'safe'. We have also disabled ISR to run in IRAM to allow more freedom, but I guess that did not work out in your case. Really, if you follow my advice above, you will not have the issues you are seeing now.
